Source/MQTTnet/Client/MqttClientOptionsBuilder.cs

```diff
@@ -86,33 +112,69 @@ public MqttClientOptionsBuilder WithWebSocketServer(string uri)
+        }
+
+        public MqttClientOptionsBuilder WithTls(SslProtocols sslProtocol,
+            Func<X509Certificate, X509Chain, SslPolicyErrors, IMqttClientOptions, bool> certificateValidationCallback = null)
+        {
+            return WithTls(new byte[][] { }, sslProtocol, certificateValidationCallback);
+        }
+
+        public MqttClientOptionsBuilder WithTls(byte[][] certificates,
+            SslProtocols sslProtocol = SslProtocols.Tls12,
+            Func<X509Certificate, X509Chain, SslPolicyErrors, IMqttClientOptions, bool> certificateValidationCallback = null)
+        {
+            return WithTls(false, certificates, sslProtocol, certificateValidationCallback);
+        }
+
+        public MqttClientOptionsBuilder WithTls(bool ignoreCertificateRevocationErrors,
+            byte[][] certificates = null,
+            SslProtocols sslProtocol = SslProtocols.Tls12,
+            Func<X509Certificate, X509Chain, SslPolicyErrors, IMqttClientOptions, bool> certificateValidationCallback = null)
+        {
+            return WithTls(false, ignoreCertificateRevocationErrors, certificates, sslProtocol, certificateValidationCallback);
+        }
+
+        public MqttClientOptionsBuilder WithTls(bool ignoreCertificateChainErrors,
+            bool ignoreCertificateRevocationErrors = false,
+            byte[][] certificates = null,
+            SslProtocols sslProtocol = SslProtocols.Tls12,
+            Func<X509Certificate, X509Chain, SslPolicyErrors, IMqttClientOptions, bool> certificateValidationCallback = null)
+        {
+            return WithTls(false, ignoreCertificateChainErrors, ignoreCertificateRevocationErrors, certificates, sslProtocol, certificateValidationCallback);
+        }
+
         public MqttClientOptionsBuilder WithTls(
-            bool allowUntrustedCertificates = false,
+            bool allowUntrustedCertificates,
             bool ignoreCertificateChainErrors = false,
             bool ignoreCertificateRevocationErrors = false,
-            params byte[][] certificates)
+            byte[][] certificates = null,
+            SslProtocols sslProtocol = SslProtocols.Tls12,
+            Func<X509Certificate, X509Chain, SslPolicyErrors, IMqttClientOptions, bool> certificateValidationCallback = null)
         {
             _tlsOptions = new MqttClientTlsOptions
             {
                 UseTls = true,
                 AllowUntrustedCertificates = allowUntrustedCertificates,
                 IgnoreCertificateChainErrors = ignoreCertificateChainErrors,
                 IgnoreCertificateRevocationErrors = ignoreCertificateRevocationErrors,
-                Certificates = certificates?.ToList()
+                Certificates = certificates?.ToList(),
+                SslProtocol = sslProtocol,
+                CertificateValidationCallback = certificateValidationCallback
             };
 
             return this;
         }
 
-        public MqttClientOptionsBuilder WithTls()
```

Comment:
Isn't this a breaking change when being removed? The other methods require at least one parameter 'allowUntrustedCertificates'.  Or did I overlook something?

Reply:
You are right, but if you make the method to use a custom object instead of parameters then this problem will go away.